source/develop/release-management/features/storage/4k-blocks.md

```diff
@@ -0,0 +1,165 @@
+---
+title: 4k drive support
+category: feature
+authors: dchaplyg
+feature_name: 4k drive support
+feature_modules: vdsm
+feature_status: Planning
+---
+
+# 4k drive support
+
+# Summary
+
+VDSM assumes, that 512 bytes is a minimally accessible block on the storage device, which can be written or read atomically. That was true for a long time, but now most of the storage devices changed their block size to the 4096 bytes (4 kilobytes). It is still compatible with VDSM, cause while we are requesting 512 atomic read or write, underlying storage will atomically read 4096 bytes, throw away 3584 bytes and return just 512. As it still acts same way, it has no impact on the VDSM. 
+
+Unfortunately, some 4k block storage devices either refuse to read/write smaller blocks atomically (some SAN appliances) or have a performance impact while emulating 512 bytes block device (like VDO). 
+
+Therefore without 4k block device support oVirt either can't use those devices as a storage domain or will have performance degradation on those domains.
+
+## Owner
+
+*   Name: [Denis Chaplygin](https://github.com/akashihi)
+*   Email: <dchaplyg@redhat.com>
+
+## Current status
+
+*   Target Release: 4.3
+*   Status: Prototyping
+*   Last updated: August 22, 2018
+
+# Implementation details
+
+VDSM heavily relies on 512 block support and it is hardcoded to a lots of places.
+
+The general approach is to refactor all block size dependent code to use block size from the storage domain, while keeping it hardcoded to 512 bytes in it and in a second step, implement block size detection at the SD level and replace hardcoded value with actual value.
+
+Subsystem specific notes are below:
+
+## Sanlock
+
+Sanlock is used by oVirt for locking and fencing shared resources. Latest released sanlock have native support for 4k devices, but it comes with price of great disk space overhead. 
+
+Sanlock is planned to be extended with new on-disk formats, incluging:
+
+ - ALIGN1M | SECTOR512: max_hosts 2000
+ - ALIGN1M | SECTOR4K:  max_hosts 250
+ - ALIGN2M | SECTOR4K:  max_hosts 500
+ - ALIGN4M | SECTOR4K:  max_hosts 1000
+ - ALIGN8M | SECTOR4K:  max_hosts 2000
+
+Thus VDSM may use ALIGN1M | SECTOR512 format for 512 bytes devices, consuming 1MB for each sanlock file. On the 4k block devices VDSM could use ALIGN1M | SECTOR4K or ALIGN2M | SECTOR4K, thus consuming either 1MB or 2MB for each sanlock file. In case of ALIGN2M | SECTOR4K usage, all lockspace files will be bigger, comparing to 512 bytes format. 
+
+With patch [1] it will be possible, to specify desired alignment, but it is not yet available in the python bindings (needs to be implemented).
+
+## Blocksize detection
+
+There is no reliable way to detect size of the logical block of the underlying device in case of a network filesystem. Both NFS (v3, v4, v4.1, v4.2) and Ceph have no aligment restrictions enforced, therefore they can't be probed for a mimally allowed block IO size. 
```

Comment:
But we can detect reliably with GlusterFS, so maybe we should take advantage of this to provide better user experience when using this storage type.

Also, we can detect reliably if we cannot detect - if we can 1 byte direct I/O request, it means the storage capabilities cannot be detect, and the user must configure the block size. If the storage support detection, we  can tell the user what are the storage capabilities.